src/tree.rs

```diff
@@ -126,12 +128,156 @@ pub fn minimum_spanning_tree(
+    for leaf_node in spanning_tree_graph.node_indices() {
+        if spanning_tree_graph.neighbors(leaf_node).count() == 1 {
+            node_queue.push_back(leaf_node);
+        }
+        same_partition_tracker[leaf_node.index()].push(leaf_node.index());
+    }
+
+    // BFS search for balanced nodes
+    let mut balanced_nodes: Vec<(usize, Vec<usize>)> = vec![];
+    let mut seen_nodes = HashSet::with_capacity(spanning_tree_graph.node_count());
+    while !node_queue.is_empty() {
+        let node = node_queue.pop_front().unwrap();
+        let pop = pops[node.index()];
+
+        // Mark as seen; push to queue if only one unseen neighbor
+        let unseen_neighbors: Vec<NodeIndex> = spanning_tree
+            .graph
+            .neighbors(node)
+            .filter(|node| !seen_nodes.contains(&node.index()))
+            .collect();
+
+        if unseen_neighbors.len() == 1 {
+            // This will be false if root
+            let neighbor = unseen_neighbors[0];
+            pops[neighbor.index()] += pop;
+            let mut current_partition_tracker = same_partition_tracker[node.index()].clone();
+            same_partition_tracker[neighbor.index()].append(&mut current_partition_tracker);
+
+            if !node_queue.contains(&neighbor) {
+                node_queue.push_back(neighbor);
+            }
+        } else if unseen_neighbors.is_empty() {
+            // root
+            break;
+        } else {
+            // Not at the leaves of the unseen subgraph
+            continue;
+        }
+
+        // Check if balanced
+        if pop >= pop_target * (1.0 - epsilon) && pop <= pop_target * (1.0 + epsilon) {
+            balanced_nodes.push((node.index(), same_partition_tracker[node.index()].clone()));
+        }
+
+        seen_nodes.insert(node.index());
+    }
+
+    balanced_nodes
+}
+
+/// Bipartition graph into two contiguous, population-balanced components using
+/// mst. Assumes that the graph is contiguous. See :func:`~bipartition_tree` for
+/// details on how balance is defined.
+///
+/// :param PyGraph graph: Undirected graph
+/// :param weight_fn: A callable object (function, lambda, etc) which
+///     will be passed the edge object and expected to return a ``float``. See
+///     :func:`~minimum_spanning_tree` for details.
+/// :param pops: The populations assigned to each node in the graph.
+/// :param float pop_target: The population target to reach when partitioning
+///     the graph.
+/// :param float epsilon: The maximum percent deviation from the pop_target
+///     allowed while still being a valid balanced cut edge.
+///
+/// :returns: A list of tuples, with each tuple representing a distinct
+///     balanced edge that can be cut. The tuple contains the root of one of the
+///     two partitioned subtrees and the set of nodes making up that subtree.
+#[pyfunction]
+#[pyo3(text_signature = "(graph, weight_fn, pop, target_pop, epsilon)")]
+pub fn bipartition_graph_mst(
```

Comment:
IMO this function can be omitted from our API since it's just calling `minimum_spanning_tree` and `bipartion_tree` but we already provide these functions as part of our API.

Reply:
I disagree. This allows reuse of the spanning tree object if there are no balanced edges detected and reduces the memory allocs (i.e. the amount of times the graph is cloned to create a new spanning tree object).

Reply:
That's a fair point but I think we can avoid compromising performance and still omit this function with a different design here. We can implement a new pyclass `SpanningTreeSampler` that generates random spanning trees and reuse internally the same memory for stroring the MST and avoid unnecessary allocations. Users can then call:
```python
sampler = retworkx.SpanningTreeSampler(graph)
balanced_nodes = []
while not balanced_nodes:
    tree = sampler.sample()
    balanced_nodes = retworkx.bipartition_tree(tree, ..)
```
to replicate the output of `bipartition_graph_mst` (and it'll be marginally slower).

My main motivation for the above design is two-fold:
1. Sampling a  spanning tree is an interesting problem on its own with different algorithms in the literature (which we can implement at a later point) and more users will benefit from it.
2. The output of `bipartition_graph_mst` feels a bit artificial to me, e.g cutting an output edge will not necessarily cut the input graph into two connected components but it'll only cut the tree that we randomly drew.